<h1>Исследование надежности заемщиков<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Постановка-задачи." data-toc-modified-id="Постановка-задачи.-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>Постановка задачи.</a></span><ul class="toc-item"><li><span><a href="#Описание-данных" data-toc-modified-id="Описание-данных-1.1"><span class="toc-item-num">1.1&nbsp;&nbsp;</span>Описание данных</a></span></li></ul></li><li><span><a href="#Ознакомление-с-данными" data-toc-modified-id="Ознакомление-с-данными-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>Ознакомление с данными</a></span><ul class="toc-item"><li><span><a href="#Создадим-датафрейм-из-файла-данных.-Выведем-первыее-20-строк." data-toc-modified-id="Создадим-датафрейм-из-файла-данных.-Выведем-первыее-20-строк.-2.1"><span class="toc-item-num">2.1&nbsp;&nbsp;</span>Создадим датафрейм из файла данных. Выведем первыее 20 строк.</a></span></li><li><span><a href="#Получим-информацию-о-датафрейме." data-toc-modified-id="Получим-информацию-о-датафрейме.-2.2"><span class="toc-item-num">2.2&nbsp;&nbsp;</span>Получим информацию о датафрейме.</a></span></li></ul></li><li><span><a href="#Предобработка-данных" data-toc-modified-id="Предобработка-данных-3"><span class="toc-item-num">3&nbsp;&nbsp;</span>Предобработка данных</a></span><ul class="toc-item"><li><span><a href="#Посчитаем-количество-пропущенных-значений-для-каждого-столбца." data-toc-modified-id="Посчитаем-количество-пропущенных-значений-для-каждого-столбца.-3.1"><span class="toc-item-num">3.1&nbsp;&nbsp;</span>Посчитаем количество пропущенных значений для каждого столбца.</a></span></li><li><span><a href="#Заполним-пропуски-в--столбце-total_income" data-toc-modified-id="Заполним-пропуски-в--столбце-total_income-3.2"><span class="toc-item-num">3.2&nbsp;&nbsp;</span>Заполним пропуски в  столбце total_income</a></span></li><li><span><a href="#Обработка-аномальных-значений" data-toc-modified-id="Обработка-аномальных-значений-3.3"><span class="toc-item-num">3.3&nbsp;&nbsp;</span>Обработка аномальных значений</a></span></li></ul></li><li><span><a href="#Изменение-типов-данных" data-toc-modified-id="Изменение-типов-данных-4"><span class="toc-item-num">4&nbsp;&nbsp;</span>Изменение типов данных</a></span></li><li><span><a href="#Обработка-дубликатов." data-toc-modified-id="Обработка-дубликатов.-5"><span class="toc-item-num">5&nbsp;&nbsp;</span>Обработка дубликатов.</a></span></li><li><span><a href="#Категоризация-данных" data-toc-modified-id="Категоризация-данных-6"><span class="toc-item-num">6&nbsp;&nbsp;</span>Категоризация данных</a></span></li><li><span><a href="#Исследование-целевых-зависимостей" data-toc-modified-id="Исследование-целевых-зависимостей-7"><span class="toc-item-num">7&nbsp;&nbsp;</span>Исследование целевых зависимостей</a></span><ul class="toc-item"><li><span><a href="#Есть-ли-зависимость-между-количеством-детей-и-возвратом-кредита-в-срок?" data-toc-modified-id="Есть-ли-зависимость-между-количеством-детей-и-возвратом-кредита-в-срок?-7.1"><span class="toc-item-num">7.1&nbsp;&nbsp;</span>Есть ли зависимость между количеством детей и возвратом кредита в срок?</a></span><ul class="toc-item"><li><span><a href="#Вывод:" data-toc-modified-id="Вывод:-7.1.1"><span class="toc-item-num">7.1.1&nbsp;&nbsp;</span>Вывод:</a></span></li></ul></li><li><span><a href="#Есть-ли-зависимость-между-семейным-положением-и-возвратом-кредита-в-срок?" data-toc-modified-id="Есть-ли-зависимость-между-семейным-положением-и-возвратом-кредита-в-срок?-7.2"><span class="toc-item-num">7.2&nbsp;&nbsp;</span>Есть ли зависимость между семейным положением и возвратом кредита в срок?</a></span><ul class="toc-item"><li><span><a href="#Вывод:" data-toc-modified-id="Вывод:-7.2.1"><span class="toc-item-num">7.2.1&nbsp;&nbsp;</span>Вывод:</a></span></li></ul></li><li><span><a href="#Есть-ли-зависимость-между-уровнем-дохода-и-возвратом-кредита-в-срок?" data-toc-modified-id="Есть-ли-зависимость-между-уровнем-дохода-и-возвратом-кредита-в-срок?-7.3"><span class="toc-item-num">7.3&nbsp;&nbsp;</span>Есть ли зависимость между уровнем дохода и возвратом кредита в срок?</a></span><ul class="toc-item"><li><span><a href="#Вывод:" data-toc-modified-id="Вывод:-7.3.1"><span class="toc-item-num">7.3.1&nbsp;&nbsp;</span>Вывод:</a></span></li></ul></li><li><span><a href="#Как-разные-цели-кредита-влияют-на-его-возврат-в-срок?" data-toc-modified-id="Как-разные-цели-кредита-влияют-на-его-возврат-в-срок?-7.4"><span class="toc-item-num">7.4&nbsp;&nbsp;</span>Как разные цели кредита влияют на его возврат в срок?</a></span></li></ul></li><li><span><a href="#Заключение.-Итоговая-таблица" data-toc-modified-id="Заключение.-Итоговая-таблица-8"><span class="toc-item-num">8&nbsp;&nbsp;</span>Заключение. Итоговая таблица</a></span></li><li><span><a href="#Выводы." data-toc-modified-id="Выводы.-9"><span class="toc-item-num">9&nbsp;&nbsp;</span>Выводы.</a></span></li></ul></div>

## Постановка задачи.

Заказчик — кредитный отдел банка. Нужно разобраться, влияет ли семейное положение, количество детей клиента, уровень дохода и цель кредита на факт погашения кредита в срок. Входные данные от банка — статистика о платёжеспособности клиентов.
Результаты исследования будут учтены при построении модели кредитного скоринга — специальной системы, которая оценивает способность потенциального заёмщика вернуть кредит банку.

### Описание данных

Описание данных
- children — количество детей в семье
- days_employed — общий трудовой стаж в днях
- dob_years — возраст клиента в годах
- education — уровень образования клиента
- education_id — идентификатор уровня образования
- family_status — семейное положение
- family_status_id — идентификатор семейного положения
- gender — пол клиента
- income_type — тип занятости
- debt — имел ли задолженность по возврату кредитов
- total_income — ежемесячный доход
- purpose — цель получения кредита

##  Ознакомление с данными

###   Создадим датафрейм из файла данных. Выведем первыее 20 строк.

In [45]:
import pandas as pd

try:
    data = pd.read_csv('data.csv')
except:
    data = pd.read_csv('/datasets/data.csv')
data.head(20)

children  days_employed  dob_years            education  education_id  \
0          1   -8437.673028         42               высшее             0   
1          1   -4024.803754         36              среднее             1   
2          0   -5623.422610         33              Среднее             1   
3          3   -4124.747207         32              среднее             1   
4          0  340266.072047         53              среднее             1   
5          0    -926.185831         27               высшее             0   
6          0   -2879.202052         43               высшее             0   
7          0    -152.779569         50              СРЕДНЕЕ             1   
8          2   -6929.865299         35               ВЫСШЕЕ             0   
9          0   -2188.756445         41              среднее             1   
10         2   -4171.483647         36               высшее             0   
11         0    -792.701887         40              среднее             1   
12         0            NaN         65              среднее             1   
13         0   -1846.641941         54  неоконченное высшее             2   
14         0   -1844.956182         56               высшее             0   
15         1    -972.364419         26              среднее             1   
16         0   -1719.934226         35              среднее             1   
17         0   -2369.999720         33               высшее             0   
18         0  400281.136913         53              среднее             1   
19         0  -10038.818549         48              СРЕДНЕЕ             1   

       family_status  family_status_id gender income_type  debt  \
0    женат / замужем                 0      F   сотрудник     0   
1    женат / замужем                 0      F   сотрудник     0   
2    женат / замужем                 0      M   сотрудник     0   
3    женат / замужем                 0      M   сотрудник     0   
4   гражданский брак                 1      F   пенсионер     0   
5   гражданский брак                 1      M   компаньон     0   
6    женат / замужем                 0      F   компаньон     0   
7    женат / замужем                 0      M   сотрудник     0   
8   гражданский брак                 1      F   сотрудник     0   
9    женат / замужем                 0      M   сотрудник     0   
10   женат / замужем                 0      M   компаньон     0   
11   женат / замужем                 0      F   сотрудник     0   
12  гражданский брак                 1      M   пенсионер     0   
13   женат / замужем                 0      F   сотрудник     0   
14  гражданский брак                 1      F   компаньон     1   
15   женат / замужем                 0      F   сотрудник     0   
16   женат / замужем                 0      F   сотрудник     0   
17  гражданский брак                 1      M   сотрудник     0   
18    вдовец / вдова                 2      F   пенсионер     0   
19         в разводе                 3      F   сотрудник     0   

     total_income                                 purpose  
0   253875.639453                           покупка жилья  
1   112080.014102                 приобретение автомобиля  
2   145885.952297                           покупка жилья  
3   267628.550329              дополнительное образование  
4   158616.077870                         сыграть свадьбу  
5   255763.565419                           покупка жилья  
6   240525.971920                       операции с жильем  
7   135823.934197                             образование  
8    95856.832424                   на проведение свадьбы  
9   144425.938277                 покупка жилья для семьи  
10  113943.491460                    покупка недвижимости  
11   77069.234271       покупка коммерческой недвижимости  
12            NaN                         сыграть свадьбу  
13  130458.228857                 приобретение автомобиля  
14  165127.911772              покупка жилой недвижимости  
15  116820.904450 

###  Получим информацию о датафрейме.

In [46]:
display(data.info())
data.describe()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          21525 non-null  int64  
 1   days_employed     19351 non-null  float64
 2   dob_years         21525 non-null  int64  
 3   education         21525 non-null  object 
 4   education_id      21525 non-null  int64  
 5   family_status     21525 non-null  object 
 6   family_status_id  21525 non-null  int64  
 7   gender            21525 non-null  object 
 8   income_type       21525 non-null  object 
 9   debt              21525 non-null  int64  
 10  total_income      19351 non-null  float64
 11  purpose           21525 non-null  object 
dtypes: float64(2), int64(5), object(5)
memory usage: 2.0+ MB


None

children  days_employed     dob_years  education_id  \
count  21525.000000   19351.000000  21525.000000  21525.000000   
mean       0.538908   63046.497661     43.293380      0.817236   
std        1.381587  140827.311974     12.574584      0.548138   
min       -1.000000  -18388.949901      0.000000      0.000000   
25%        0.000000   -2747.423625     33.000000      1.000000   
50%        0.000000   -1203.369529     42.000000      1.000000   
75%        1.000000    -291.095954     53.000000      1.000000   
max       20.000000  401755.400475     75.000000      4.000000   

       family_status_id          debt  total_income  
count      21525.000000  21525.000000  1.935100e+04  
mean           0.972544      0.080883  1.674223e+05  
std            1.420324      0.272661  1.029716e+05  
min            0.000000      0.000000  2.066726e+04  
25%            0.000000      0.000000  1.030532e+05  
50%            0.000000      0.000000  1.450179e+05  
75%            1.000000      0.000000  2.034351e+05  
max            4.000000      1.000000  2.265604e+06

Выводы: Колонка **children**, **days_employed dob_years** содержит аномальные значения. Колонки **days_employed** и **total_income содержа**т пропуски значений. 

## Предобработка данных

### Посчитаем количество пропущенных значений для каждого столбца.

In [47]:
(data.isna().sum())

children               0
days_employed       2174
dob_years              0
education              0
education_id           0
family_status          0
family_status_id       0
gender                 0
income_type            0
debt                   0
total_income        2174
purpose                0
dtype: int64

### Заполним пропуски в  столбце total_income

Заполним пропуски в  **столбце total_income** медианным значением по каждому типу из столбца income_type. В данном случае медиана лучшее решение, т.к. позволит не учитывать артефактно большие и малые значения.

In [48]:
for i in data['income_type'].unique():
    data.loc[(data['income_type'] == i) & (data['total_income'].isna()), 'total_income'] = \
    data.loc[(data['income_type'] == i), 'total_income'].median()

проверим результат

In [49]:
data['total_income'].isna().sum()

0

### Обработка аномальных значений

Заменим  отрицательные значения days_employed  значения положительными с помощью метода abs().


In [50]:
data['days_employed'] = data['days_employed'].abs()

Для каждого типа занятости выведем медианное значение трудового стажа days_employed в днях.

In [51]:
data.groupby('income_type')['days_employed'].median()

income_type
безработный        366413.652744
в декрете            3296.759962
госслужащий          2689.368353
компаньон            1547.382223
пенсионер          365213.306266
предприниматель       520.848083
сотрудник            1574.202821
студент               578.751554
Name: days_employed, dtype: float64

У двух типов (безработные и пенсионеры) получатся аномально большие значения. Исправить такие значения сложно, поэтому придется оставить их как есть.

Выведем перечень уникальных значений столбца children.

In [52]:
data['children'].unique()

array([ 1,  0,  3,  2, -1,  4, 20,  5], dtype=int64)

В столбце children есть два аномальных значения. Удалим строки, в которых встречаются такие аномальные значения из датафрейма data.

In [53]:
data = data[(data['children'] != -1) & (data['children'] != 20)]

Проверим результат

In [54]:
data['children'].unique()

array([1, 0, 3, 2, 4, 5], dtype=int64)

Заполним пропуски в столбце days_employed медианными значениями по каждому типу занятости income_type

In [55]:
for i in data['income_type'].unique():
    data.loc[(data['income_type'] == i) & (data['days_employed'].isna()), 'days_employed'] = \
    data.loc[(data['income_type'] == i), 'days_employed'].median()

Убедимся, что пропусков больше нет

In [56]:
data.isna().sum()

children            0
days_employed       0
dob_years           0
education           0
education_id        0
family_status       0
family_status_id    0
gender              0
income_type         0
debt                0
total_income        0
purpose             0
dtype: int64

## Изменение типов данных

Заменим вещественный тип данных в столбце total_income на целочисленный с помощью метода astype().

In [57]:
data['total_income']= data['total_income'].astype(int)

## Обработка дубликатов.

Выведем уникальные значения столбца education

In [58]:
data['education'].unique()

array(['высшее', 'среднее', 'Среднее', 'СРЕДНЕЕ', 'ВЫСШЕЕ',
       'неоконченное высшее', 'начальное', 'Высшее',
       'НЕОКОНЧЕННОЕ ВЫСШЕЕ', 'Неоконченное высшее', 'НАЧАЛЬНОЕ',
       'Начальное', 'Ученая степень', 'УЧЕНАЯ СТЕПЕНЬ', 'ученая степень'],
      dtype=object)

Обработаем неявные дубликаты в столбце education. В этом столбце есть одни и те же значения, но записанные по-разному: с использованием заглавных и строчных букв. Приведем их к нижнему регистру.

In [59]:
data['education'] = data['education'].str.lower()

проверим:

In [60]:
data['education'].unique()

array(['высшее', 'среднее', 'неоконченное высшее', 'начальное',
       'ученая степень'], dtype=object)

Сосчитаем дубликаты:

In [61]:
data.duplicated().sum()

71

Удалим дубликаты:

In [62]:
data = data.drop_duplicates( keep='first')

проверим:

In [63]:
data.duplicated().sum()

0

## Категоризация данных

На основании диапазонов, указанных ниже, создим в датафрейме data столбец total_income_category с категориями:

- 0–30000 — 'E';
- 30001–50000 — 'D';
- 50001–200000 — 'C';
- 200001–1000000 — 'B';
- 1000001 и выше — 'A'.

создадим функцию categorize_income()

In [64]:
def categorize_income(total_income):
    
    if (total_income >0) & (total_income <=30000 ):        
        return 'E'
    if (total_income >=30001) & (total_income <=50000 ):
        return 'D'
    if (total_income >=50001) & (total_income <=200000 ):
        return 'C'
    if (total_income >=200001) & (total_income <=1000000  ):
        return 'B'
    if total_income >1000001:
        return 'A'
    

применим функцию с помощью метода apply()

In [65]:
data['total_income_category'] = data['total_income'].apply(categorize_income)

выведем на экран перечень уникальных целей взятия кредита из столбца purpose.

In [66]:
data['purpose'].unique()

array(['покупка жилья', 'приобретение автомобиля',
       'дополнительное образование', 'сыграть свадьбу',
       'операции с жильем', 'образование', 'на проведение свадьбы',
       'покупка жилья для семьи', 'покупка недвижимости',
       'покупка коммерческой недвижимости', 'покупка жилой недвижимости',
       'строительство собственной недвижимости', 'недвижимость',
       'строительство недвижимости', 'на покупку подержанного автомобиля',
       'на покупку своего автомобиля',
       'операции с коммерческой недвижимостью',
       'строительство жилой недвижимости', 'жилье',
       'операции со своей недвижимостью', 'автомобили',
       'заняться образованием', 'сделка с подержанным автомобилем',
       'получение образования', 'автомобиль', 'свадьба',
       'получение дополнительного образования', 'покупка своего жилья',
       'операции с недвижимостью', 'получение высшего образования',
       'свой автомобиль', 'сделка с автомобилем',
       'профильное образование', 'высшее об

Создадим функцию categorize_purpose(), которая на основании данных из столбца purpose сформирует новый столбец purpose_category, в который войдут следующие категории:
- 'операции с автомобилем',
- 'операции с недвижимостью',
- 'проведение свадьбы',
- 'получение образования'.

In [67]:
# тут можно применить и конструкцию try-except, но по моему менению sring.find по сути и сам такой конструкцией является.
def categorize_purpose(s):
    if s.find('автом')>= 0:
        return 'операции с автомобилем'
    if s.find('жил')>= 0  or s.find('движ')>= 0 :
        return 'операции с недвижимостью'
    if s.find('свад')>= 0 : 
        return 'проведение свадьбы'
    if s.find('образ')>= 0 :
        return 'получение образования'
    else:
        return 'категоря не определена'

применим функцию методом apply()

In [68]:
data['purpose_category'] = data['purpose'].apply(categorize_purpose)

проверим результат:

In [69]:
data[data['purpose_category'] == 'категоря не определена'].sum()

children                 0.0
days_employed            0.0
dob_years                0.0
education                0.0
education_id             0.0
family_status            0.0
family_status_id         0.0
gender                   0.0
income_type              0.0
debt                     0.0
total_income             0.0
purpose                  0.0
total_income_category    0.0
purpose_category         0.0
dtype: float64

## Исследование целевых зависимостей

### Есть ли зависимость между количеством детей и возвратом кредита в срок?

Создадим таблицу, где по категорям количества детей посчитаем количество заемщиков, количество должников, долю должников.

In [70]:
data.groupby('children')['debt'].agg(['count','sum','mean']).sort_values(by='mean', ascending=False)

count   sum      mean
children                       
4            41     4  0.097561
2          2052   194  0.094542
1          4808   444  0.092346
3           330    27  0.081818
0         14091  1063  0.075438
5             9     0  0.000000

Так как количество заемщиков с детьми =3, 4, 5 слишком малочисленнно с сравнении с категориями где их 1 и 2
то предлагаю их объединить в одну, категории с детьми =1 и =2 сопоставимы по уровню задолженности поэтому их тоже объединим.
для этого напишем фунуцию.

In [71]:
def children_category(row):
    if row['children'] == 0:
        return 'нет детей'
    elif 1 <= row['children'] <= 2:
        return '1-2 ребенка'
    else:
        return '3 и более детей'
data['children_category'] = data.apply(children_category, axis=1)
#data.head(15)


Создадим таблицу, где по категорям количества детей посчитаем количество заемщиков, количество должников, долю должников и % должников для удобства визуального восприятия.

In [72]:
debt_children = data.groupby('children_category')['debt'].agg(['count','sum','mean']).sort_values(by='mean', ascending=True)
def part_to_percent (part):
    return "{0:.2%}".format(part)
debt_children['% должников'] = debt_children['mean'].apply(part_to_percent)
debt_children = debt_children.rename(columns={'count': 'Количество заемщиков', 'sum':'Количество должников', 'mean':'Доля должников'})
debt_children

Количество заемщиков  Количество должников  Доля должников  \
children_category                                                               
нет детей                         14091                  1063        0.075438   
3 и более детей                     380                    31        0.081579   
1-2 ребенка                        6860                   638        0.093003   

                  % должников  
children_category              
нет детей               7.54%  
3 и более детей         8.16%  
1-2 ребенка             9.30%

#### Вывод:

Наименьшее количество должников категория без детей, что может быть объяснено их большей финансовой свободой.
На 2м месте категория с количеством детей >=3 объяснений предложить не могу, возможно проблема в количестве выборки = 380
замыкают ряд заёмшики с количеством детей 1 и 2.

Проведем статистические анализ для категориальных признаков хи-квадрат по кростабам  для  выборок 
с 0 и 1 детьми 
с 0 и 2
и с 1 и 2 

In [73]:
#Создадим такие выборки
ch_0_1 = data[(data['children'] == 0) | (data['children'] == 1)]
ch_0_2 = data[(data['children'] == 0) | (data['children'] == 2)]
ch_1_2 = data[(data['children'] == 1) | (data['children'] == 2)]
from scipy.stats import chi2_contingency, fisher_exact
print ('Результат теста хи-квадрат для сравнения выборок 0 и 1 детей:',chi2_contingency(pd.crosstab(ch_0_1['children'], ch_0_1['debt'])))
print ('Результат теста хи-квадрат для сравнения выборок 0 и 2 детей:',chi2_contingency(pd.crosstab(ch_0_2['children'], ch_0_2['debt'])))
print ('Результат теста хи-квадрат для сравнения выборок 1 и 2 детей:',chi2_contingency(pd.crosstab(ch_1_2['children'], ch_1_2['debt'])))

Результат теста хи-квадрат для сравнения выборок 0 и 1 детей: (13.73615068616518, 0.0002103661683480545, 1, array([[12967.38832742,  1123.61167258],
       [ 4424.61167258,   383.38832742]]))
Результат теста хи-квадрат для сравнения выборок 0 и 2 детей: (8.839726420391974, 0.002947433247616135, 1, array([[12993.78219662,  1097.21780338],
       [ 1892.21780338,   159.78219662]]))
Результат теста хи-квадрат для сравнения выборок 1 и 2 детей: (0.058236533471726484, 0.8093051985894079, 1, array([[4360.84198251,  447.15801749],
       [1861.15801749,  190.84198251]]))


Итог: то статистически достоверное отличие есть между категорями как с 0 и 1 так и с 0 и 2 детьми
а вот категории с 1м и с 2мя детьми не отличаются.

### Есть ли зависимость между семейным положением и возвратом кредита в срок?

In [74]:
debt_fam_status = data.groupby('family_status')['debt'].agg(['count','sum','mean']).sort_values(by='mean', ascending=True)
debt_fam_status['% должников'] = debt_fam_status['mean'].apply(part_to_percent)
debt_fam_status = debt_fam_status.rename(columns={'count': 'Количество заемщиков', 'sum':'Количество должников', 'mean':'Доля должников'})
debt_fam_status

Количество заемщиков  Количество должников  \
family_status                                                       
вдовец / вдова                          951                    63   
в разводе                              1189                    84   
женат / замужем                       12261                   927   
гражданский брак                       4134                   385   
Не женат / не замужем                  2796                   273   

                       Доля должников % должников  
family_status                                      
вдовец / вдова               0.066246       6.62%  
в разводе                    0.070648       7.06%  
женат / замужем              0.075606       7.56%  
гражданский брак             0.093130       9.31%  
Не женат / не замужем        0.097639       9.76%

#### Вывод:

Вывод: на основе полученной таблицы можем заключить, что наилучшие заемщики - 'катеория вдовец / вдова', немного им уступает категория 'в разводе'Вероятно опть же по причине бОльшей финансовой свободы.
На 3м месте категория  'женат / замужем'
Сравнительно наихудшие 'Не женат / не замужем'.
странно что отличаются категории 'женат / замужем'  и 'гражданский брак', хотя последняя в приерно в 3 раза меньше и возможно причина в этом.

### Есть ли зависимость между уровнем дохода и возвратом кредита в срок?

Создадим сводную аналогичную таблиццу.

In [75]:
debt_total_inc_cat = data.groupby('total_income_category')['debt'].agg(['count','sum','mean']).sort_values(by='mean', ascending=True)
debt_total_inc_cat['% должников'] = debt_total_inc_cat['mean'].apply(part_to_percent)
debt_total_inc_cat = debt_total_inc_cat.rename(columns={'count': 'Количество заемщиков', 'sum':'Количество должников', 'mean':'Доля должников'})
debt_total_inc_cat

Количество заемщиков  Количество должников  \
total_income_category                                               
D                                       349                    21   
B                                      5014                   354   
A                                        25                     2   
C                                     15921                  1353   
E                                        22                     2   

                       Доля должников % должников  
total_income_category                              
D                            0.060172       6.02%  
B                            0.070602       7.06%  
A                            0.080000       8.00%  
C                            0.084982       8.50%  
E                            0.090909       9.09%

#### Вывод:

Категории А и Е настолько малочисленны, что выводы по ним делать не стоит.
Посмотрим сопоставимые по представительности  категории В - условно выше стреднего и С - условно средний доход.
Люди с средним доходом чаще попадают в категорию должников.


In [76]:
#уточним достоветность отличия категорий B и C 
b_c = data[(data['total_income_category'] == 'B') | (data['total_income_category'] == 'C')]
print ('Результат теста хи-квадрат для сравнения выборок B и С категоий дохода:',chi2_contingency(pd.crosstab(b_c['total_income_category'], b_c['debt'])))

Результат теста хи-квадрат для сравнения выборок B и С категоий дохода: (10.33732175796385, 0.001303670356415118, 1, array([[ 4605.16799618,   408.83200382],
       [14622.83200382,  1298.16799618]]))


Категории дохода В и С статистически значимо отличаются по доле просроченных кредитов.

### Как разные цели кредита влияют на его возврат в срок?

In [77]:
debt_purpose = data.groupby('purpose_category')['debt'].agg(['count','sum','mean']).sort_values(by='mean', ascending=True)
debt_purpose['% должников'] = debt_purpose['mean'].apply(part_to_percent)
debt_purpose = debt_purpose.rename(columns={'count': 'Количество заемщиков', 'sum':'Количество должников', 'mean':'Доля должников'})
debt_purpose

Количество заемщиков  Количество должников  \
purpose_category                                                       
операции с недвижимостью                 10751                   780   
проведение свадьбы                        2313                   183   
получение образования                     3988                   369   
операции с автомобилем                    4279                   400   

                          Доля должников % должников  
purpose_category                                      
операции с недвижимостью        0.072551       7.26%  
проведение свадьбы              0.079118       7.91%  
получение образования           0.092528       9.25%  
операции с автомобилем          0.093480       9.35%

Жилье это главное, похожим образом дело обстоит и со свадьбами, но порядок  расходов на эти цели несопоставим, наверно поэтому
кредиты на свадьба сравнительно легче вернуть. Получение образование и операции с автомобилям замыкают ряд надежности заемщиков.


## Заключение. Итоговая таблица

С помощью метода df.pivot_table создадим общую сводную таблицу, где группировка будет происходить по всем 4м исследуемым факторам одновремено. Применим порог размера выборки >30 это спорный ход признаю и от этой границы зависят выводы

In [78]:
data_pivot = data.pivot_table(index=['children_category', 'total_income_category', 'family_status', 'purpose_category'],  values='debt', aggfunc = ('mean', lambda X:
                                                                       X.count()))

data_pivot = data_pivot.rename(columns={'<lambda_0>':'К-во наблюдений','mean': 'Доля должников'})

#применим отсечку, чтобы не учитывать малочисленные выборки, это очень условное действие, и наверно не самое правильное....
data_pivot = data_pivot[data_pivot['К-во наблюдений'] >30] 
sorted_data_pivot = data_pivot.sort_values(by=['Доля должников'], ascending=False)

display(sorted_data_pivot.info())
display (sorted_data_pivot.head(5))
(sorted_data_pivot.tail(5))

<class 'pandas.core.frame.DataFrame'>
MultiIndex: 58 entries, ('1-2 ребенка', 'C', 'гражданский брак', 'получение образования') to ('нет детей', 'B', 'в разводе', 'операции с автомобилем')
Data columns (total 2 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   К-во наблюдений  58 non-null     float64
 1   Доля должников   58 non-null     float64
dtypes: float64(2)
memory usage: 1.5+ KB


None

К-во наблюдений  \
children_category total_income_category family_status         purpose_category                            
1-2 ребенка       C                     гражданский брак      получение образования                99.0   
                                        Не женат / не замужем операции с автомобилем               89.0   
                                        гражданский брак      операции с автомобилем              105.0   
нет детей         C                     гражданский брак      получение образования               195.0   
1-2 ребенка       C                     гражданский брак      операции с недвижимостью            236.0   

                                                                                        Доля должников  
children_category total_income_category family_status         purpose_category                          
1-2 ребенка       C                     гражданский брак      получение образования           0.202020  
                                        Не женат / не замужем операции с автомобилем          0.179775  
                                        гражданский брак      операции с автомобилем          0.161905  
нет детей         C                     гражданский брак      получение образования           0.148718  
1-2 ребенка       C                     гражданский брак      операции с недвижимостью        0.127119

К-во наблюдений  \
children_category total_income_category family_status    purpose_category                            
нет детей         D                     женат / замужем  операции с недвижимостью             81.0   
1-2 ребенка       B                     гражданский брак проведение свадьбы                  183.0   
нет детей         D                     женат / замужем  получение образования                34.0   
1-2 ребенка       D                     женат / замужем  операции с недвижимостью             36.0   
нет детей         B                     в разводе        операции с автомобилем               49.0   

                                                                                   Доля должников  
children_category total_income_category family_status    purpose_category                          
нет детей         D                     женат / замужем  операции с недвижимостью        0.037037  
1-2 ребенка       B                     гражданский брак проведение свадьбы              0.032787  
нет детей         D                     женат / замужем  получение образования           0.029412  
1-2 ребенка       D                     женат / замужем  операции с недвижимостью        0.027778  
нет детей         B                     в разводе        операции с автомобилем          0.020408

 по таблице: получено 58 комбинаций исследуемых факторов с отсечкой по количеству наблюдений >30.

Самые благонадежные заемщики:
- Доля просрочки - 2%
- Детей - нет 
- Категория дохода - выше среднего.
- Семейное положение -  в разводе
- Цель кредита - операции с автомобилем


Наименее благонадежные:
- Доля просрочки - 20%
- Детей - 1-2 
- Категория дохода - средний.
- Семейное положение -  гражданский брак
- Цель кредита - получение образования 

##  Выводы.

Исходные датасет содержит 21525 записей и 12 колонок.
Данные подвергнуты предобработке на основе следующих проблем - содержали пропуски значений, дубликаты и аномальные значения:

- Колонки days_employed и total_income содержили пропуски 2174 значений, что около 10% датасета.Помимо этого days_employed содержал и артефактные отрицательыне значения.

- 2174 отсутствующих значений days_employed были заполнены медианным значением по каждому типу из столбца income_type. Медианная характеристика лучше средней т.к. позволяет не учитывать аномальные значения по краям последовательности. 

- days_employed содержали отрицательные значения, что вероятнее всего опечатки, значения - скорректированы методом abs().

- Аномальные значения трудового стажа для безработных и пенсионеров более 365000дней ~ 1000 лет не представилось возможным скорректировать. Но и особого смысла в этом нет, т.к. данные сведения не являются исследуеемым параметром.

- Отсутствующие данные total_income  были заполнены также медианным значением по каждому типу из столбца income_type. тип данных данной колонки был заменен с float на int методом .astype(int)

- Количество детей children содержали значения = -1 и =20 Данные строки были удалены.

- Данные колонки education содержали регистрозависимые дубликаты. Приведены к нижнему регистру методом .str.lower() Дубликаты удалены методом .drop_duplicates( keep='first')

**В пунтках 7.* показано, что все исследуемые параметры имеют влияние на целевую функцию возврата кредита в срок.** 

**Общая формулировка - надежность заемщика убывает в ряду признаков**:
- **Детей**: нет, 1-2, 3 и более.
- **Семейное положение**:состоящие или состоявшие в браке заемщики, гражданский брак, не женат/не замужем.
- **Категория дохода**: условно ниже среднего(но выборка несопоставимо меньше B с С),условно выше среднего B, условно средний C.
- **Цель кредита**: операции с недвижимостью, проведение свадьбы, получение образования, операции с автомобилем.

Появление большого количества пропусков в колонке total_income  требует отдельного разбирательства т.к было установлено, что данный параметр заемщика имеет влияние на количество просрочки. Это ли сбой выгрузки или что-то еще...

Общая рекомендация провести ревизию интрефейса загрузки данных в БД:
- 1. На предмет наличия фильтров проверяющих вводимые данные на обязательность важных полей как например total_income 
- 2. На предмет наличия фильтров на проверку адекватности значений(например чтобы не допускать количества детей -1 или трудовой стаж в 1000 лет) 
- 3. Заполнение категории цель кредита лучше сделать в форме выбора значения из словаря стандарных целей.

Дать совет насчет отбора лучших признаков для прогнозирования помог бы факторный анализ или главных компонент, но это еще предстоит изучить....

 